Fix custom generate from local directory

manueldeprada opened the pull request:
Passing a local folder to `custom_generate` via `model.generate(**inputs, custom_generate="custom_generate_example", ...)` fails because 
1. `shutil.copy` in `src/transformers/dynamic_module_utils.py` silently fails if the parent directory structure doesn't exist. 
2. The relative imports do not get correctly copied to the fresh submodule in .cache.

So, this PR does:
1. Create parent dirs before copying files (custom_generate dir)
2. Correctly copy relative import files to the submodule folder.
3. Update docs.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_38916). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

manueldeprada committed:
Fix custom generate from local directory:
1. Create parent dirs before copying files (custom_generate dir)
2. Correctly copy relative imports to the submodule file.
3. Update docs.

gante reviewed (approved):
LGTM :)

Missing: a test. E.g. we can add a `custom_generate` folder in `tests/fixtures` and test we can run it in `GenerationIntegrationTests`
